Mathlib/Tactic/Relation/Rfl.lean

```diff
@@ -0,0 +1,63 @@
+import Lean
+import Mathlib.Tactic.Relation.Basic
+
+namespace Mathlib.Tactic.Relation.Rfl
+
+open Lean Meta
+
+/-- Environment extensions for `refl` lemmas -/
+initialize reflExtension : SimpleScopedEnvExtension (Name × Array DiscrTree.Key)
+(DiscrTree Name) ← registerSimpleScopedEnvExtension {
+    name := `refl
+    addEntry := fun dt (n, ks) => dt.insertCore ks n
+    initial := {}
+  }
+
+/-- add reflexivity attribute if valid -/
+def reflAttr : AttributeImpl where
+  name := `refl
+  descr := "reflexivity relation"
+  add decl _ kind := do
+    MetaM.run' do
+      let declTy := (← getConstInfo decl).type
+      let (xs, _, targetTy) ← withReducible <| forallMetaTelescopeReducing declTy
+      if xs.size < 1 then
+        throwError "@[refl] attribute only applies to lemmas proving x ∼ x, got {declTy} with
+        too few arguments"
+      else
+        match ← relationAppM? targetTy with
+        | some (rel, lhs, rhs) =>
+          unless (← isDefEq lhs rhs) do
+            throwError "@[refl] attribute only applies to lemmas proving x ∼ x, got {declTy}
+            with {lhs} ~ {rhs} instead"
+          let key ← withReducible <| DiscrTree.mkPath rel
+          reflExtension.add (decl, key) kind
+        | none =>
+          throwError "@[refl] attribute only applies to lemmas proving x ∼ x, got {declTy}"
+
+initialize registerBuiltinAttribute reflAttr
+
+
+/-- look up reflexivity lemmas -/
+def reflLemmas (env : Environment) : DiscrTree Name :=
+  reflExtension.getState env
+
+open Lean.Elab.Tactic in
+
+/-- This tactic applies to a goal whose target has the form `x ~ x`, where `~` is a reflexive
+relation, that is, a relation which has a reflexive lemma tagged with the attribute [refl].
+-/
+elab "rfl" : tactic =>
```

Comment:
Ah, this is the problem. Using `elab "rfl"` means you are declaring another syntax called `rfl`, which takes priority over the other one and all its implementations. You want to use ``elab_rules : tactic | `(rfl) => ...`` instead here to make your implementation play well with the others.